packages/core/src/sync-realtime/service.ts

```diff
@@ -202,444 +219,506 @@ export class RealtimeSyncService extends Emittery<RealtimeSyncEvents> {
         service: "realtime",
-        msg: `Error while fetching latest block (error=${`${error.name}: ${error.message}`})`,
+        msg: `Realtime sync task failed (network=${
+          this.network.name
+        }, error=${`${error.name}: ${error.message}`})`,
+        network: this.network.name,
       });
-    }
-  };
-
-  private buildQueue = () => {
-    const queue = createQueue<RealtimeBlockTask>({
-      worker: async ({ task }) => {
-        await this.blockTaskWorker(task);
-      },
-      options: { concurrency: 1, autoStart: false },
-      onError: ({ error, task }) => {
-        if (this.isShuttingDown) return;
-
-        this.common.logger.warn({
-          service: "realtime",
-          msg: `Realtime sync task failed (network=${
-            this.network.name
-          }, error=${`${error.name}: ${error.message}`})`,
-          network: this.network.name,
-          hash: task.block.hash,
-          parentHash: task.block.parentHash,
-          number: task.block.number,
-          timestamp: task.block.timestamp,
-        });
+      // Note: Good spot to throw a fatal error.
+    } finally {
+      this.isProcessingBlock = false;
 
-        // Default to a retry (uses the retry options passed to the queue).
-        queue.addTask({ ...task });
-      },
-    });
-
-    return queue;
+      if (this.isProcessBlockQueued) {
+        this.isProcessBlockQueued = false;
+        await this.processBlock();
+      }
+    }
   };
 
-  private blockTaskWorker = async ({ block }: RealtimeBlockTask) => {
-    const previousHeadBlock = this.blocks[this.blocks.length - 1];
-
-    // If no block is passed, fetch the latest block.
-    const newBlockWithTransactions = block;
-    const newBlock = rpcBlockToLightBlock(newBlockWithTransactions);
-
-    // 1) We already saw and handled this block. No-op.
-    if (this.blocks.find((b) => b.hash === newBlock.hash)) {
+  private blockHandler = async (newBlock: RealtimeBlock) => {
+    // We already saw and handled this block. No-op.
+    if (this.mostRecentBlock.hash === newBlock.hash) {
       this.common.logger.trace({
         service: "realtime",
-        msg: `Already processed block at ${newBlock.number} (network=${this.network.name})`,
+        msg: `Already processed block at ${hexToNumber(
+          newBlock.number,
+        )} (network=${this.network.name})`,
       });
+
       return;
     }
 
-    // 2) This is the new head block (happy path). Yay!
+    const sync = this.determineSyncPath(newBlock);
+    if (sync === "traverse") await this.syncTraverse(newBlock);
+    else await this.syncBatch(newBlock);
+
+    // If this block moves the finality checkpoint, remove now-finalized blocks from the local chain
+    // and mark data as cached in the store.
     if (
-      newBlock.number === previousHeadBlock.number + 1 &&
-      newBlock.parentHash === previousHeadBlock.hash
+      hexToNumber(newBlock.number) >
+      this.finalizedBlockNumber + 2 * this.network.finalityBlockCount
```

Comment:
If we do choose to use the `"finalized"` tag, the `finalityBlockCount` logic would also need to be dynamic (I think).